README

 IMPORTANT:
   This code is still pre-beta and has not been audited for flaws that
   could introduce weaknesses in the encryption. Please don't use it for
   any confidential data before the beta release.

------------------------------------------------------------------------
 STATUS:
  See http://wiki.debian.org/DebianInstaller/PartmanCrypto
  and also TODO and BUGS in the source tree

------------------------------------------------------------------------
The scripts do stuff more or less in the following order:

 1. choose_method/crypto
 2. active_partition/*
 3. choose_partition/crypto
 4. init.d/crypto
 5. blockdev-keygen
 6. update.d/crypto_visuals
 7. finish.d/check_randomfskey, aptinstall_crypto, fstab_loop

choose_method/
  crypto

    Entry in the "Use as:" menu. All encryption types (loop-AES,
    dm-crypt) are merged into one choice. Writes "crypto"
    into $dev/$id/method and activates the below crypto options.

active_partition/*

    Entries in the partition menu. Which are shown depends on crypt_type
    and cipher. loop-AES shows no keysize option because it's ciphers have
    fixed keysizes. The chosen values are stored in $dev/$id/<entry>

choose_partition/
  crypto

    "Setup encrypted volumes" in the main menu. Checks for valid cipher
    options and safe swap. Creates keyfiles, erases to-be-encrypted
    parititons and finally does losetup/dmsetup.

init.d/
  crypto

    Creates the partman cryptdisks once for each crypto partition.
  
update.d/
  crypto_visuals

    Shows "crypt" in the method column, and the name of the crypt disk as 
    mountpoint, so the user can see which loop/dm device is which after
    they have been setup.

finish.d/
  crypto_check_mountpoints

    Shows a warning if user chose to create a filesystem on a crypt disk
    with keytype random. User can force to continue, maybe that should 
    eventually be disabled. Note: random keytype can make sense for the
    /tmp partition, but actually requires random/1777 keytype.
   
finish.d/
  crypto_config
    
    Replaces loop devices in fstab with corresponding real devices and 
    adds crypto mount options. This is a bit of a hack used because we 
    can't use get_mountoptions from partman. Crypto settings are stored 
    in dev/$id of cleartext partition, but everything else is stored in
    the crypt_device dev/$id. We want the mount options, finish.d checks,
    etc. handled by whatever feels responsible for what should be store
    d on the encrypted partition, so this seems necessary.

finish.d/
  crypto_aptinstall

    Installs userspace tools and helper packages for encryption. This 
    is responsible for installing loop-aes-utils and loop-aes-rootfs (TODO)
    into the target system. Similarily for dm-crypt, installs cryptsetup 
    etc.

blockdev-keygen
   
    Asks for a passphrase, confirms that it has been entered correctly 
    and optionally tests for a minimum length. If the name for a keyfile
    is specified as third argument it creates a gnupg keyfile for use
    with loop-AES, otherwise it just echoes the pass to stdout. 
    
    The actual encryption keys for loop-AES keyfiles are generated from
    /dev/random, so we need plenty of entropy. This script uses the 
    cdebconf plugin "entropy" if available. See notes at the beginning
    of the script for important steps for the caller.

finish-install.d/
  05crypto

    Run by finish-install.d hook; Checks for loop-AES partitions,
    determines target kernel version from base-installer/kernel/image
    and installs a corresponding loop-aes-$KVERS package into /target.

ciphers/$crypt_type/$cipher

    Lists supported ciphers for each crypt_type. The following files
    exist for each cipher directory. They are used to display a short
    description of the cipher for the user and to determine allowed 
    crypto parameters (TODO):

    - desc (description)
    - options (crypt_type specific)

    Structure is
      ciphers/$crypt_type/$cipher/<file>
  
mountoptions
mountoptions/crypto

    Should be no longer required. Maybe resurrect once we find a cleaner
    approach for setting crypto-specific mount options.



debian/changelog
partman-crypto (66) UNRELEASED; urgency=low

  [ Thiemo Nagel ]
  * Use the same security setting for wiping existing volumes
    than previous settings, when the volume was already encrypted.
    Closes: #726448

 -- Christian Perrier <bubulle@debian.org>  Sat, 09 Nov 2013 09:12:37 +0100

partman-crypto (65) unstable; urgency=low

  [ Updated translations ]
  * Czech (cs.po) by Miroslav Kure
  * French (fr.po) by Christian Perrier
  * Japanese (ja.po) by Kenshi Muto
  * Simplified Chinese (zh_CN.po) by YunQiang Su

 -- Christian Perrier <bubulle@debian.org>  Wed, 06 Nov 2013 06:49:03 +0100

partman-crypto (64) unstable; urgency=low

  [ Thiemo Nagel ]
  * Make "Erasing data" step optional. blockdev-wipe is too slow
    Closes: #722898
  * blockdev-wipe: Set blocksize to 512k
  * blockdev-wipe: Sync at most once per second
  * blockdev-wipe: Allocate buffer from heap instead of stack
  * blockdev-wipe: Reduce progress indicator granularity to 1/1000

  [ Updated translations ]
  * Polish (pl.po) by Michał Kułach

 -- Christian Perrier <bubulle@debian.org>  Wed, 30 Oct 2013 20:24:43 +0100

partman-crypto (63) unstable; urgency=low

  [ Thiemo Nagel ]
  * remove broken BLKGETSIZE fallback from blockdev-wipe. Closes: #722696

 -- Christian Perrier <bubulle@debian.org>  Wed, 02 Oct 2013 07:26:54 +0200

partman-crypto (61) unstable; urgency=low

  [ Updated translations ]
  * Tajik (tg.po) by Victor Ibragimov

 -- Christian Perrier <bubulle@debian.org>  Sun, 08 Sep 2013 19:33:44 +0200

partman-crypto (60) unstable; urgency=low

  [ Dmitrijs Ledkovs ]
  * Guard with apt-install '|| true' in case apt-install exists with error
    code 1, indicating delayed installation.
  * Set debian source format to '3.0 (native)'.

 -- Christian Perrier <bubulle@debian.org>  Sat, 13 Jul 2013 14:26:27 +0200

partman-crypto (59) unstable; urgency=low

  * Add myself to uploads.
  * Bump debhelper compat to 9.
  * Use standard VCS URLs.
  * Move unsafe swap checks from commit.d to check.d to be executed
    earlier and thus returning quicker back to partitioning menu
  * Install cryptsetup user space utilities, earlier, in finish.d instead
    of post-base-installer.d. This way cryptsetup utilities are installed
    if post-base-installer is not executed at all (like in ubiquity) or
    didn't manage to run for any other reason. This is also more inline
    with other fs-specific utilities and other partman-* packages.

 -- Dmitrijs Ledkovs <dmitrijs.ledkovs@canonical.com>  Thu, 23 May 2013 12:05:59 +0100

partman-crypto (58) unstable; urgency=low

  [ Dmitrijs Ledkovs ]
  * Remove remaining loop-aes code.

  [ Colin Watson ]
  * Remove redundant and unused debian/install file.
  * Use correct compiler and strip when cross-building.

  [ Updated translations ]
  * Croatian (hr.po) by Tomislav Krznar

 -- Christian Perrier <bubulle@debian.org>  Thu, 16 May 2013 18:42:44 +0200

partman-crypto (57) unstable; urgency=low

  [ Updated translations ]
  * Croatian (hr.po) by Tomislav Krznar

 -- Christian Perrier <bubulle@debian.org>  Sat, 30 Mar 2013 17:39:01 +0100

partman-crypto (56) unstable; urgency=low

  [ Updated translations ]
  * Catalan (ca.po) by Jordi Mallach

 -- Christian Perrier <bubulle@debian.org>  Wed, 12 Dec 2012 07:15:01 +0100

partman-crypto (55) unstable; urgency=low

  * Add myself to Uploaders.

  [ Updated translations ]
  * Asturian (ast.po) by ivarela
  * Spanish (es.po) by Javier Fernández-Sanguino
  * Ukrainian (uk.po) by Yuri Chornoivan

 -- Christian Perrier <bubulle@debian.org>  Tue, 16 Oct 2012 12:37:34 +0200

partman-crypto (54) unstable; urgency=low

  [ Bastian Blank ]
  * Disable loop-aes support. (closes: #680943)
  * Support and use xts-plain64 by default. (closes: #482092)
    - Double key size if xts mode is selected.

 -- Julien Cristau <jcristau@debian.org>  Fri, 03 Aug 2012 10:36:03 +0200

partman-crypto (53) unstable; urgency=low

  * Team upload

  [ Updated translations ]
  * Croatian (hr.po) by Tomislav Krznar
  * Portuguese (Brazil) (pt_BR.po) by Felipe Augusto van de Wiel (faw)

 -- Christian Perrier <bubulle@debian.org>  Sun, 24 Jun 2012 08:27:54 +0200

partman-crypto (52) unstable; urgency=low

  * Team upload

  [ Updated translations ]
  * Galician (gl.po) by Jorge Barreiro

 -- Christian Perrier <bubulle@debian.org>  Sun, 17 Jun 2012 12:51:26 +0200

partman-crypto (51) unstable; urgency=low

  * Team upload
  * Replace XC-Package-Type with Package-Type

  [ Updated translations ]
  * Tibetan (bo.po) by Tennom
  * Welsh (cy.po) by Dafydd Tomos
  * Galician (gl.po) by Jorge Barreiro
  * Lithuanian (lt.po) by Rimas Kudelis
  * Latvian (lv.po) by Rūdolfs Mazurs
  * Macedonian (mk.po) by Arangel Angov
  * Panjabi (pa.po) by A S Alam
  * Portuguese (Brazil) (pt_BR.po) by Felipe Augusto van de Wiel (faw)
  * Traditional Chinese (zh_TW.po) by Yao Wei (魏銘廷)

 -- Christian Perrier <bubulle@debian.org>  Sat, 16 Jun 2012 07:30:33 +0200

partman-crypto (50) unstable; urgency=low

  [ Updated translations ]
  * Asturian (ast.po) by Mikel González
  * Bulgarian (bg.po) by Damyan Ivanov
  * Danish (da.po) by Joe Hansen
  * Estonian (et.po) by Mattias Põldaru
  * Indonesian (id.po) by Mahyuddin Susanto
  * Icelandic (is.po) by Sveinn í Felli
  * Kannada (kn.po) by Prabodh C P
  * Dutch (nl.po) by Jeroen Schot

 -- Otavio Salvador <otavio@debian.org>  Thu, 15 Mar 2012 14:59:17 -0300

partman-crypto (49) unstable; urgency=low

  * Merge from Ubuntu:
    - Allow crypto kernel modules to be built-in.
  * The fix for #642115 was actually working around a 'dmsetup table' output
    format regression in libdevmapper 1.02.65 / lvm2 2.02.86, fixed in
    libdevmapper 1.02.66 / lvm2 2.02.87.  For robustness, handle both old
    and new output formats (closes: #647540, LP: #892903).

  [ Updated translations ]
  * Hebrew (he.po) by Lior Kaplan
  * Kazakh (kk.po) by Baurzhan Muftakhidinov
  * Kannada (kn.po) by vignesh prabhu
  * Polish (pl.po) by Marcin Owsiany


base64/base64.c
/*
 * This is a minimal base64 encoder. It is derived from b64.c
 * version 08/04/01 -- Create Version 0.00.00B by Bob Trower,
 * with the following changes by Max Vozeler <xam@debian.org>
 * 
 * - Stripped code not required for encoding
 * - Use LF line endings instead of CRLF
 * - Use fixed LINESIZE of 60 to follow uuencode -m
 *
 * Copyright (c) 2001 Bob Trower, Trantor Standard Systems Inc.
 * 
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the
 * Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall
 * be included in all copies or substantial portions of the
 * Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

#include <stdio.h>

#define LINESIZE 60

static const char cb64[]=
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

/*
** encodeblock
**
** encode 3 8-bit binary bytes as 4 '6-bit' characters
*/

void encodeblock( unsigned char in[3], unsigned char out[4], int len )
{
    out[0] = cb64[ in[0] >> 2 ];
    out[1] = cb64[ ((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4) ];
    out[2] = (unsigned char) (len > 1 ? cb64[ ((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6) ] : '=');
    out[3] = (unsigned char) (len > 2 ? cb64[ in[2] & 0x3f ] : '=');
}

/*
** encode
**
** base64 encode a stream adding padding and line breaks as per spec.
*/

void encode(FILE *infile, FILE *outfile)
{
    unsigned char in[3], out[4];
    int i, len, blocksout = 0;

    while( !feof( infile ) ) {
        len = 0;
        for( i = 0; i < 3; i++ ) {
            in[i] = (unsigned char) getc( infile );
            if( !feof( infile ) ) {
                len++;
            }
            else {
                in[i] = 0;
            }
        }
        if( len ) {
            encodeblock( in, out, len );
            for( i = 0; i < 4; i++ ) {
                putc( out[i], outfile );
            }
            blocksout++;
        }
        if( blocksout >= (LINESIZE/4) || feof( infile ) ) {
            if( blocksout ) {
                fprintf( outfile, "\n" );
            }
            blocksout = 0;
        }
    }
}

int main(int argc, char **argv)
{
    encode(stdin, stdout);
    if (ferror(stdin) || ferror(stdout))
        return 1;
    return 0;
}


lib/crypto-base.sh
. /lib/partman/lib/base.sh
. /lib/partman/lib/commit.sh

# Would this partition be allowed as a physical volume for crypto?
crypto_allowed() {
	local dev=$1
	local id=$2

	# Allow unless this is a crypto device
	[ ! -f "$dev/crypt_realdev" ]
}

crypto_list_allowed() {
	partman_list_allowed crypto_allowed
}

crypto_list_allowed_free() {
	local line

	IFS="$NL"
	for line in $(crypto_list_allowed); do
		restore_ifs
		local dev="${line%%$TAB*}"
		local rest="${line#*$TAB}"
		local id="${rest%%$TAB*}"
		if [ -e "$dev/locked" ] || [ -e "$dev/$id/locked" ]; then
			continue
		fi
		echo "$line"
		IFS="$NL"
	done
	restore_ifs
}

# Prepare a partition for use as a physical volume for encryption. If this
# returns true, then it did some work and a commit is necessary. Prints the
# new path.
crypto_prepare () {
	local dev="$1"
	local id="$2"
	local num size parttype fs path

	cd "$dev"
	open_dialog PARTITION_INFO "$id"
	read_line num id size freetype fs path x7
	close_dialog

	if [ "$fs" = free ]; then
		local newtype

		case $freetype in
		    primary)
			newtype=primary
			;;
		    logical)
			newtype=logical
			;;
		    pri/log)
			local parttype
			open_dialog PARTITIONS
			while { read_line x1 x2 x3 parttype x5 x6 x7; [ "$parttype" ]; }; do
				if [ "$parttype" = primary ]; then
					has_primary=yes
				fi
			done
			close_dialog
			if [ "$has_primary" = yes ]; then
				newtype=logical
			else
				newtype=primary
			fi
			;;
		esac

		open_dialog NEW_PARTITION $newtype ext2 $id full $size
		read_line num id x3 x4 x5 path x7
		close_dialog
	fi

	mkdir -p "$id"
	local method="$(cat "$id/method" 2>/dev/null || true)"
	if [ "$method" = swap ]; then
		disable_swap "$dev" "$id"
	fi
	if [ "$method" != crypto ]; then
		crypto_prepare_method "$id" dm-crypt || return 1
		rm -f "$id/use_filesystem"
		rm -f "$id/format"
		echo dm-crypt >"$id/crypto_type"
		echo crypto >"$id/method"

		while true; do
			local code=0
			ask_active_partition "$dev" "$id" "$num" || code=$?
			if [ "$code" -ge 128 ] && [ "$code" -lt 192 ]; then
				exit "$code" # killed by signal
			elif [ "$code" -ge 100 ]; then
				break
			fi
		done

		update_partition "$dev" "$id"
		echo "$path"
		return 0
	fi

	echo "$path"
	return 1
}

dm_dev_is_safe() {
	local maj min dminfo deps
	maj="$1"
	min="$2"

	# First try the device itself
	dminfo=$(dmsetup table -j$maj -m$min 2> /dev/null | \
		 head -n1 | sed 's/^[^ ]*: //' | cut -d' ' -f3) || return 1
	if [ "$dminfo" = crypt ]; then
		return 0
	fi

	# Then check its deps instead
	deps=$(dmsetup deps -j "$maj" -m "$min" 2> /dev/null) || return 1
	deps=$(echo "$deps" | sed -e 's/.*://;s/[ (]//g;s/)/ /g')

	# deps is now a list like 3,2 3,1
	for dep in $deps; do
		maj=${dep%%,*}
		min=${dep##*,}
		dm_dev_is_safe "$maj" "$min" || return 1
	done

	return 0
}

dm_is_safe() {
	# Might be non-encrypted, e.g. LVM2
	local dminfo major minor
	type dmsetup > /dev/null 2>&1 || return 1

	dminfo=$(dmsetup info -c "$1" | tail -1) || return 1
	major=$(echo "$dminfo" | sed 's/ \+/ /g' | cut -d' ' -f2)
	minor=$(echo "$dminfo" | sed 's/ \+/ /g' | cut -d' ' -f3)

	dm_dev_is_safe "$major" "$minor" || return 1
	return 0
}

swap_is_safe () {
	local swap
	local IFS="
"

	for swap in $(cat /proc/swaps); do
		case $swap in
		    Filename*)
			continue
			;;
		    /dev/mapper/*)
			dm_is_safe ${swap%% *} || return 1
			;;
		    *)
			# Presume not safe
			return 1
			;;
		esac
	done

	return 0
}

get_free_mapping() {
	for n in 0 1 2 3 4 5 6 7; do
		if [ ! -b "/dev/mapper/crypt$n" ]; then
			echo "crypt$n"
			break
		fi
	done
}

setup_dmcrypt () {
	local mapping device cipher iv hash size pass
	mapping=$1
	device=$2
	cipher=$3
	iv=$4
	hash=$5
	size=$6
	pass=$7

	[ -x /sbin/cryptsetup ] || return 1

	# xts modes needs double the key size
	[ "${iv%xts-*}" = "${iv}" ] || size="$(($size * 2))"

	log-output -t partman-crypto \
	/sbin/cryptsetup -c $cipher-$iv -d $pass -h $hash -s $size create $mapping $device
	if [ $? -ne 0 ]; then
		log "cryptsetup failed"
		return 2
	fi

	return 0
}

setup_luks () {
	local mapping device cipher iv size pass
	mapping=$1
	device=$2
	cipher=$3
	iv=$4
	size=$5
	pass=$6

	[ -x /sbin/cryptsetup ] || return 1

	# xts modes needs double the key size
	[ "${iv%xts-*}" = "${iv}" ] || size="$(($size * 2))"

	log-output -t partman-crypto \
	/sbin/cryptsetup -c $cipher-$iv -s $size luksFormat $device $pass
	if [ $? -ne 0 ]; then
		log "luksFormat failed"
		return 2
	fi

	log-output -t partman-crypto \
	/sbin/cryptsetup -d $pass luksOpen $device $mapping
	if [ $? -ne 0 ]; then
		log "luksOpen failed"
		return 2
	fi

	return 0
}

setup_cryptdev () {
	local type id realdev cryptdev
	type=$1
	id=$2
	realdev=$3

	for opt in keytype cipher keyfile ivalgorithm keyhash keysize; do
		eval local $opt

		if [ -r "$id/$opt" ]; then
			eval $opt=$(cat $id/$opt)
		else
			eval $opt=""
		fi
	done

	case $type in
	    dm-crypt)
		cryptdev=$(mapdevfs $realdev)
		cryptdev="${cryptdev##*/}_crypt"
		if [ -b "/dev/mapper/$cryptdev" ]; then
			cryptdev=$(get_free_mapping)
			if [ -z "$cryptdev" ]; then
				return 1
			fi
		fi
		if [ $keytype = passphrase ]; then
			setup_luks $cryptdev $realdev $cipher $ivalgorithm $keysize $keyfile || return 1
		elif [ $keytype = random ]; then
			setup_dmcrypt $cryptdev $realdev $cipher $ivalgorithm plain $keysize /dev/urandom || return 1
		else
			setup_dmcrypt $cryptdev $realdev $cipher $ivalgorithm $keyhash $keysize $keyfile || return 1
		fi
		cryptdev="/dev/mapper/$cryptdev"
		;;

	esac

	echo $cryptdev > $id/crypt_active
	db_subst partman-crypto/text/in_use DEV "${cryptdev##*/}"
	db_metaget partman-crypto/text/in_use description
	partman_lock_unit $(mapdevfs $realdev) "$RET"
	return 0
}

crypto_do_wipe () {
	local template dev fifo pid x
	template=$1
	dev=$2
	fifo=/var/run/wipe_progress

	mknod $fifo p
	/bin/blockdev-wipe -s $((512*1024)) $dev > $fifo &
	pid=$!

	cancelled=0
	db_capb backup align progresscancel
	db_progress START 0 1000 ${template}_title
	db_progress INFO ${template}_text
	while read x <&9; do
		db_progress STEP 1
		if [ $? -eq 30 ]; then
			cancelled=1
			kill $pid
			break
		fi
	done 9< $fifo
	db_progress STOP
	db_capb backup align

	rm $fifo
	wait $pid
	ret=$?

	[ $cancelled -eq 1 ] && ret=0
	return $ret
}

crypto_wipe_device () {
	local device method interactive targetdevice
	device=$1
	method=$2
	interactive=$3
	if [ "$interactive" != no ]; then
		interactive=yes
	fi
	ret=1

	if [ -r $part/crypto_type ] && [ "$(cat $part/crypto_type)" = dm-crypt ]; then
		type=crypto
	else
		type=plain
	fi

	if [ $interactive = yes ]; then
		# Confirm before erasing
		template="partman-crypto/${type}_warn_erase"
		db_set $template false
		db_subst $template DEVICE $(humandev $device)
		db_input critical $template || true
		db_go || return
		db_get $template
		if [ "$RET" != true ]; then
			return 0
		fi
	fi

	# Setup crypto
	if [ "$type" = crypto ]; then
		targetdevice=$(get_free_mapping)
		setup_dmcrypt $targetdevice $device aes xts-plain64 plain 128 /dev/urandom || return 1
		targetdevice="/dev/mapper/$targetdevice"
	else
		# Just wipe the device with zeroes
		targetdevice=$device
	fi

	# Erase
	template="partman-crypto/progress/${type}_erase"
	db_subst ${template}_title DEVICE $(humandev $device)
	db_subst ${template}_text DEVICE $(humandev $device)
	if ! crypto_do_wipe $template $targetdevice; then
		template="partman-crypto/${type}_erase_failed"
		db_subst $template DEVICE $(humandev $device)
		db_input critical $template || true
		db_go
	else
		ret=0
	fi

	# Teardown crypto
	if [ "$type" = crypto ]; then
		log-output -t partman-crypto /sbin/cryptsetup remove ${targetdevice##/dev/mapper/}
	fi

	return $ret
}

crypto_dochoice () {
	local part type cipher option value

	part=$1
	type=$2
	cipher=$3
	option=$4

	if [ ! -f /lib/partman/ciphers/$type/$cipher/$option ] && \
	   [ ! -f /lib/partman/ciphers/$type/$option ]; then
		exit 0
	fi

	if [ -f $part/$option ]; then
		value=$(cat $part/$option)
		template="partman-crypto/text/$option/$value"
		db_metaget $template description && value="$RET"
	else
		value="none"
		template=partman-basicfilesystems/text/no_mountpoint
		db_metaget $template description && value="$RET"
	fi

	db_metaget partman-crypto/text/specify_$option description
	printf "%s\t%s\${!TAB}%s\n" "$option" "$RET" "$value"
}

crypto_dooption () {
	local part type cipher option choices altfile template

	part=$1
	type=$2
	cipher=$3
	option=$4

	if [ -f /lib/partman/ciphers/$type/$cipher/$option ]; then
		altfile="/lib/partman/ciphers/$type/$cipher/$option"
	else
		altfile="/lib/partman/ciphers/$type/$option"
	fi

	choices=$(
		for value in $(cat $altfile); do
			description="$value"
			template="partman-crypto/text/$option/$value"
			db_metaget $template description && description="$RET"
			printf "%s\t%s\n" $value "$description"
		done
	)

	template="partman-crypto/$option"
	debconf_select critical $template "$choices" "" || exit 0
	if [ "$RET" = none ]; then
		rm -f $part/$option
		return
	fi

	echo $RET > $part/$option
}

crypto_load_module() {
	local module=$1

	if [ "$module" = dm_mod ]; then
		if dmsetup version >/dev/null 2>&1; then
			return 0
		fi
		modprobe -q $module
		return $?
	elif [ "$module" = dm_crypt ]; then
		if dmsetup targets | cut -d' ' -f1 | grep -q '^crypt$'; then
			return 0
		fi
		modprobe -q $module
		return $?
	else
		if egrep -q "^(name|version) *: $module\$" /proc/crypto; then
			return 0
		fi
		modprobe -q $module
		return $?
	fi
}

# Loads all modules for a given crypto type and cipher
crypto_load_modules() {
	local type cipher moduledir modulefile module
	type=$1
	cipher=$2
	moduledir=/var/run/partman-crypto/modules

	if [ ! -d $moduledir ]; then
		mkdir -p $moduledir
	fi

	for modulefile in \
	  /lib/partman/ciphers/$type/module \
	  /lib/partman/ciphers/$type/$cipher/module; do
		[ -f $modulefile ] || continue
		for module in $(cat $modulefile); do
			if [ -f $moduledir/$module ]; then
				# Already loaded
				continue
			fi

			if crypto_load_module $module; then
				touch $moduledir/$module
			else
				rm -f $moduledir/$module
				return 1
			fi
		done
	done

	return 0
}

# Checks that we have sufficient memory to load crypto udebs
crypto_check_mem() {
	local verbose="$1"

	# A more or less arbitrary limit
	if [ $(memfree) -lt 10000 ]; then
		if [ "$verbose" != true ]; then
			return 1
		fi

		db_set partman-crypto/install_udebs_low_mem false
		db_fset partman-crypto/install_udebs_low_mem seen false
		db_input critical partman-crypto/install_udebs_low_mem
		db_go || true
		db_get partman-crypto/install_udebs_low_mem
		if [ "$RET" != true ]; then
			return 1
		fi
	fi

	return 0
}

# Loads additional crypto udebs
crypto_load_udebs() {
	local packages udebdir package
	packages="$*"
	udebdir=/var/run/partman-crypto/udebs

	if [ -z "$packages" ]; then
		return 0
	fi

	if [ ! -d $udebdir ]; then
		mkdir -p $udebdir
	fi

	local need_depmod=
	for package in $packages; do
		if [ -f $udebdir/$package ]; then
			continue
		fi

		crypto_check_mem true || return 1

		if ! anna-install $package; then
			db_fset partman-crypto/install_udebs_failure seen false
			db_input critical partman-crypto/install_udebs_failure
			db_go || true
			return 1
		fi

		touch $udebdir/$package
		need_depmod=1
	done

	if [ "$need_depmod" ]; then
		# The udeb installation run usually adds new kernel modules
		if [ -x /sbin/depmod ]; then
			depmod -a > /dev/null 2>&1 || true
		fi

		# Reset the capabilities after anna-install
		db_capb backup align
	fi

	return 0
}

# Sets the defaults for a given crypto type
crypto_set_defaults () {
	local part type
	part=$1
	type=$2

	[ -d $part ] || return 1

	case $type in
	    dm-crypt)
		echo aes > $part/cipher
		echo 256 > $part/keysize
		echo xts-plain64 > $part/ivalgorithm
		echo passphrase > $part/keytype
		echo sha256 > $part/keyhash
		;;
	esac
	return 0
}

# Does initial setup for a crypto method
crypto_prepare_method () {
	local part type package
	part=$1
	type=$2
	packages=''

	[ -d $part ] || return 1
	packages="cdebconf-$DEBIAN_FRONTEND-entropy"
	case $type in
	    dm-crypt)
		packages="$packages partman-crypto-dm"
		;;
	    *)
		return 1
		;;
	esac

	# 1A - Pull in the method package and additional dependencies
	crypto_load_udebs $packages || return 1

	# 1B - Verify that it worked
	crypto_check_required_tools $type || return 1

	# 2 - Set the defaults for the chosen type
	crypto_set_defaults $part $type || return 1

	# 3 - Also load the kernel modules needed for the chosen type/cipher
	[ -f $part/cipher ] || return 1
	crypto_load_modules $type $(cat $part/cipher) || return 1

	return 0
}

crypto_check_required_tools() {
	local tools

	tools="blockdev-keygen"
	case $1 in
	    dm-crypt)
		tools="$tools dmsetup cryptsetup"
		;;
	    *)
		return 1
	esac

	for tool in $tools; do
		if ! type $tool > /dev/null 2>&1 ; then
			db_fset partman-crypto/tools_missing seen false
			db_input critical partman-crypto/tools_missing
			db_go || true
			return 1
		fi
	done
	return 0
}

crypto_check_required_options() {
	local id type list options
	path=$1
	type=$2

	case $type in
	    dm-crypt)
		options="cipher keytype keyhash ivalgorithm keysize"
		;;
	esac

	list=""
	for opt in $options; do
		[ -f $path/$opt ] && continue
		db_metaget partman-crypto/text/specify_$opt description || RET="$opt:"
		desc=$RET
		db_metaget partman-crypto/text/missing description || RET="missing"
		value=$RET
		if [ "$list" ]; then
			list="$list
$desc $value"
		else
			list="$desc $value"
		fi
	done

	# If list is non-empty, at least one option is missing
	if [ ! -z "$list" ]; then
		templ="partman-crypto/options_missing"
		db_fset $templ seen false
		db_subst $templ DEVICE "$(humandev $path)"
		db_subst $templ ITEMS "$list"
		db_input critical $templ
		db_go || true
		return 1
	fi
	return 0
}

crypto_check_setup() {
	crypt=
	for dev in $DEVICES/*; do
		[ -d "$dev" ] || continue
		cd $dev

		partitions=
		open_dialog PARTITIONS
		while { read_line num id size type fs path name; [ "$id" ]; }; do
			[ "$fs" != free ] || continue
			partitions="$partitions $id"
		done
		close_dialog

		for id in $partitions; do
			[ -f $id/method ] || continue
			[ -f $id/crypto_type ] || continue

			method=$(cat $id/method)
			if [ $method != crypto ]; then
				continue
			fi
			type=$(cat $id/crypto_type)
			crypt=yes

			crypto_check_required_tools $type
			crypto_check_required_options "$dev/$id" $type
		done
	done

	if [ -z "$crypt" ]; then
		db_fset partman-crypto/nothing_to_setup seen false
		db_input critical partman-crypto/nothing_to_setup
		db_go || true
		return 1
	fi
	return 0
}

crypto_setup() {
	local interactive s dev id size path methods partitions type keytype keysize
	interactive=$1
	if [ "$interactive" != no ]; then
		interactive=yes
	fi

	commit_changes partman-crypto/commit_failed || return $?

	if ! swap_is_safe; then
		db_fset partman-crypto/unsafe_swap seen false
		db_input critical partman-crypto/unsafe_swap
		db_go || true
		return 1
	fi

	# Erase crypto-backing partitions
	for dev in $DEVICES/*; do
		[ -d "$dev" ] || continue
		cd $dev

		partitions=
		open_dialog PARTITIONS
		while { read_line num id size type fs path name; [ "$id" ]; }; do
			[ "$fs" != free ] || continue
			partitions="$partitions $id,$size,$path"
		done
		close_dialog

		for part in $partitions; do
			set -- $(IFS=, && echo $part)
			id=$1
			size=$2
			path=$3

			[ -f $id/method ] || continue
			method=$(cat $id/method)
			if [ $method != crypto ]; then
				continue
			fi

			if [ -f $id/crypt_active ] || [ -f $id/skip_erase ]; then
				continue
			fi

			if ! crypto_wipe_device $path $(cat $id/crypto_type) $interactive; then
				db_fset partman-crypto/commit_failed seen false
				db_input critical partman-crypto/commit_failed
				db_go || true
				return 1
			fi
		done
	done

	# Create keys and do dmsetup
	for dev in $DEVICES/*; do
		[ -d "$dev" ] || continue
		cd $dev

		partitions=
		open_dialog PARTITIONS
		while { read_line num id size type fs path name; [ "$id" ]; }; do
			[ "$fs" != free ] || continue
			partitions="$partitions $id,$path"
		done
		close_dialog

		for part in $partitions; do
			id=${part%,*}
			path=${part#*,}

			[ -f $id/method ] || continue
			[ -f $id/crypto_type ] || continue
			[ -f $id/cipher ] || continue
			[ -f $id/keytype ] || continue

			method=$(cat $id/method)
			if [ $method != crypto ]; then
				continue
			fi

			type=$(cat $id/crypto_type)
			keytype=$(cat $id/keytype)
			cipher=$(cat $id/cipher)

			if [ $keytype = keyfile ] || [ $keytype = passphrase ]; then
				keyfile=$(mapdevfs $path | tr / _)
				keyfile="$dev/$id/${keyfile#_dev_}"

				if [ ! -f $keyfile ]; then
					if ! /bin/blockdev-keygen "$(humandev $path)" "$keytype" "$keyfile"; then
						db_fset partman-crypto/commit_failed seen false
						db_input critical partman-crypto/commit_failed
						db_go || true
						failed=1
						break
					fi
				fi

				echo $keyfile > $id/keyfile
			fi

			if [ ! -f $id/crypt_active ]; then
				log "setting up encrypted device for $path"

				if ! setup_cryptdev $type $id $path; then
					db_fset partman-crypto/commit_failed seen false
					db_input critical partman-crypto/commit_failed
					db_go || true
					failed=1
					break
				fi
			fi
		done
	done

	if [ $failed ]; then
		return 1
	fi

	stop_parted_server

	restart_partman
	return 0
}
